src/utils.rs

```diff
@@ -0,0 +1,158 @@
+use flate2::read::GzDecoder;
+use std::path::Path;
+use std::process::{Command, ExitStatus};
+use std::{fs, io};
+use tar::Archive;
+
+/// Filenames that if found in a project mean it should be treated as a Python project,
+/// and so pass this buildpack's detection phase.
+///
+/// This list is deliberately larger than just the list of supported package manager files,
+/// so that Python projects that are missing some of the required files still pass detection,
+/// allowing us to show a more detailed error message during the build phase than is possible
+/// during detect.
+const KNOWN_PYTHON_PROJECT_FILES: [&str; 9] = [
+    ".python-version",
+    "main.py",
+    "manage.py",
+    "Pipfile",
+    "poetry.lock",
+    "pyproject.toml",
+    "requirements.txt",
+    "runtime.txt",
+    "setup.py",
+];
+
+/// Returns whether the specified project directory is that of a Python project, and so
+/// should pass buildpack detection.
+pub(crate) fn is_python_project(app_dir: &Path) -> io::Result<bool> {
+    // Until `Iterator::try_find` is stabilised, this is cleaner as a for loop.
+    for filename in KNOWN_PYTHON_PROJECT_FILES {
+        if app_dir.join(filename).try_exists()? {
```

Comment:
> Errors from std::io::Error do not include the target file or what operation failed. I've been suggesting use of the fs_err crate https://docs.rs/fs-err/latest/fs_err/index.html

Yeah, I wish the stdlib included paths too, I am subscribed to (and have a thumbs up on):
https://github.com/rust-lang/rfcs/issues/2885
https://github.com/rust-lang/rust/pull/86826

...and had seen your usage of the `fs_err` crate in the Ruby CNB and checked it out, since it looks interesting.

> though camino is another popular alternative 
> ...
> It looks like camino has this method implemented https://docs.rs/camino/1.1.3/camino/struct.Utf8Path.html#method.try_exists

I don't see how Camino solves the filename problem? It uses the stdlib `io::Error`?
https://docs.rs/camino/1.1.3/src/camino/lib.rs.html#1335

If we wanted to switch wholesale to camino, then I think this should start within libcnb.rs first.

> I think this gives a sub-optimal error message

I considered this when implementing. I tried very hard to break an app/build in order to write an integration test that tests the detection IO error message. However, I couldn't find a way to do so, since all the obvious approaches (such as `chmod`ing a file locally to remove permissions) caused Pack / `lifecycle` itself to fail, and never got as far as the buildpack. 

As such, I'm confident that this detect IO error will always only ever be due to a platform issue (eg the whole EBS mount is messed up), rather than an application source issue (eg wrong permissions on say `requirements.txt`). Therefore, having the filename be listed here doesn't add much value.

If you can demonstrate a single integration testcase that proves this is not the case, then I will happily reconsider :-)

> If they have one bad and one good file, we still fail the build as this try ? will fail eagerly. 

Yup, this is intentional. We shouldn't silently ignore errors here.

> I think we should actually not care about if the file is valid or not here at all, but always return a successful detection, then deliver a better error message in the build phase.

Pack already shows the full detect error if the build fails, so we don't need to defer to the build phase for this.

Reply:
> I don't see how Camino solves the filename problem? It uses the stdlib io::Error?

Sorry. I was parroting some community advice. I heard about both at the same time and was under the impression that camino gave better error messages, but either I misunderstood, or it's not applied to all contexts. I would need to actually exercise this code path to make sure they're not providing that functionality in some sneaky way i'm just not seeing.

> If you can demonstrate a single integration testcase that proves this is not the case, then I will happily reconsider :-)

Did you only try changing permissions or also symlinks? Here's an example of a bad symlink https://github.com/schneems/which_problem/blob/65bda424cf62bc04aa23f192e971a80e4db0ae83/which_problem/src/lib.rs#L148-L153

>  Therefore, having the filename be listed here doesn't add much value.

I think it's pretty core to the debugging process. You're operating under the assumption that this can only fail of something catastrophic happened and ALL files are failing. I'm not as confident. Even if it's not user trigerable, the first thing I would want to know if I was debugging this as a buildpack author or pack maintainer is: What file failed.

Reply:
Scratch that. I understand now. Sorry that last response was missing context on my part. The `try_exist` fails based on permissions issues, only for parent directories, not for self. It does not Err on symlinks. 

So yes, i'm confident that this would be a CNB bug and not a user issue (now). I also tried out camino and confirm it doesn't help at all with an error message in this case:

```rust
fn main() {}

#[cfg(test)]
mod tests {
    use std::ffi::OsString;

    #[test]
    fn lol() {
        let tmp_dir = tempfile::tempdir().unwrap();
        let dir = tmp_dir.path();
        let file = dir.join("lol");

        std::fs::write(&file, "lol");

        use std::os::unix::fs::PermissionsExt;
        let metadata = dir.metadata().unwrap();
        let mut perms = metadata.permissions();
        perms.set_mode(0o000);

        std::fs::set_permissions(&dir, perms).unwrap();

        camino::Utf8Path::new(file.to_str().unwrap())
            .try_exists()
            .unwrap();
    }
}
```

Gives us:

```
thread 'tests::lol' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', src/main.rs:24:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The error message raised here is:

```
        BuildpackError::DetectIo(io_error) => log_io_error(
            "Unable to complete buildpack detection",
            "determining if the Python buildpack should be run for this application",
            &io_error,
        ),
```

Which will show up as:

```
Unable to complete buildpack detection

An unexpected error occurred whilst determining if the Python buildpack should be run for this application.
            
Details: I/O Error: Permission denied
```

To an end user it would be unclear what permissions are denied (parent directory does not have read permissions) and that it's an error with pack/CNB/libcnb and not with their application.

Reply:
In the buildpack I'm using the phase "unexpected error" to cover things that are "not normal, potentially our issue but it's hard to say more than that". I have an existing TODO to explore this further:
https://github.com/heroku/buildpacks-python/blob/192230568b35c7ac32e20677e4d11838dc8b0659/src/errors.rs#L280

For now, what's there is more than enough (and a better implementation already than several of the other CNBs), and I don't see this as a blocker, so given there is a TODO I'm going to resolve this for now.

Thank you for raising some good points however - I'll bear them in mind for the future :-)